builder/simplify.cc

```diff
@@ -391,25 +425,44 @@ class simplifier : public node_mutator {
       set_result(let_stmt::make(std::move(lets), std::move(body)));
     }
   }
+  
+  void merge_loop_references(symbol_map<symbol_info> add) {
+    for (symbol_id i = 0; i < add.size(); ++i) {
+      if (!add[i]) continue;
+      std::optional<symbol_info>& info = references[i];
+      if (!info) {
+        info = std::move(add[i]);
+      } else {
+        info->ref_count += add[i]->ref_count;
+        info->referenced_in_loop = true;
+      }
+    }
+  }
 
   void visit(const loop* op) override {
+    symbol_map<symbol_info> old_references;
+    std::swap(old_references, references);
+
     interval_expr bounds = mutate(op->bounds);
     expr step = mutate(op->step);
 
     if (prove_true(bounds.min > bounds.max)) {
       // This loop is dead.
       set_result(stmt());
+      merge_loop_references(std::move(old_references));
```

Comment:
This probably shouldn't be here. If the loop is dead, we've removed these references from the program.